File: commonutils/src/main/java/net/maxsmr/commonutils/data/gson/PropertyBasedAdapter.java
package net.maxsmr.commonutils.data.gson;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.lang.reflect.Type;

import static net.maxsmr.commonutils.data.gson.GsonUtilsKt.getJsonElementAs;
import static net.maxsmr.commonutils.data.gson.GsonUtilsKt.getJsonPrimitive;

public class PropertyBasedAdapter<T> implements JsonSerializer<T>, JsonDeserializer<T> {

    private static final String CLASS_META_KEY = "CLASS_META_KEY";

    @Override
    public JsonElement serialize(T src, Type typeOfSrc, JsonSerializationContext context) {
        JsonElement jsonElem = context.serialize(src, src.getClass());
        JsonObject jsonObject = getJsonElementAs(jsonElem, JsonObject.class);
        if (jsonObject != null) {
            jsonObject.addProperty(CLASS_META_KEY, src.getClass().getCanonicalName());
        }
        return jsonElem;
    }

    @Override
    public T deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        String className = getJsonPrimitive(json, CLASS_META_KEY, String.class);
        if (className != null) {
            Class<?> clz;
            try {
                clz = Class.forName(className);
            } catch (ClassNotFoundException e) {
                throw new JsonParseException(e);
            }
            return context.deserialize(json, clz);
        }
        throw new JsonParseException(CLASS_META_KEY + " not specified");
    }

}

File: commonutils-jre/src/main/kotlin/net/maxsmr/commonutils/model/gson/PropertyBasedAdapter.kt
package net.maxsmr.commonutils.model.gson

import com.google.gson.*
import java.lang.reflect.Type

/**
 * Регистрировать для абстрактных типов;
 * при сериализации происходит дописывание его имени класса,
 * десериализация делегируется для конкретного известного класса
 *
 * ВНИМАНИЕ!
 * 1. Вероятно не применим, если у абстракции, для которой регистрируется этот адаптер, есть хотя бы 1 enum реализация.
 * В этом случае метод [serialize] для нее не вызывается, и как следствие затем при десериализации не будут найдены
 * метаданные (имя класса), что приводит к крашу.
 * 1. Можно использовать **ТОЛЬКО** для регистрации абстракцых типов - интерфейсов либо абстрактных классов. При
 * попытке зарегистрировать на конкретном типе или open классе возможна бесконечная рекурсия при сериализации (см. доку [serialize])
 */
class PropertyBasedAdapter<T: Any> : JsonSerializer<T>, JsonDeserializer<T> {

    override fun serialize(src: T, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val jsonElement = context.serialize(src, src.javaClass)
        val obj = JsonObject()
        obj.add(KEY_SOURCE, jsonElement)
        obj.addProperty(KEY_CLASS_META, src.javaClass.name)
        return obj
    }

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): T? {
        getJsonPrimitive(json, KEY_CLASS_META, String::class.java)?.let { className ->
            val clz: Class<*> = try {
                Class.forName(className)
            } catch (e: ClassNotFoundException) {
                throw JsonParseException(e)
            }
            return context.deserialize(getJsonElement(json, KEY_SOURCE, JsonElement::class.java), clz)
        }
        throw JsonParseException("$KEY_CLASS_META not specified for json: $json")
    }

    companion object {

        const val KEY_SOURCE = "SOURCE"
        const val KEY_CLASS_META = "CLASS_META"
    }
}